src/conn.c

```diff
@@ -3422,6 +3422,28 @@ natsConnection_ConnectTo(natsConnection **newConn, const char *url)
     return NATS_UPDATE_ERR_STACK(s);
 }
 
+natsStatus
+natsConnection_Reconnect(natsConnection *nc)
+{
+    if (nc == NULL)
+        return nats_setDefaultError(NATS_INVALID_ARG);
+
+    natsConn_Lock(nc);
+    if (natsConn_isClosed(nc))
+    {
+        natsConn_Unlock(nc);
+        return nats_setDefaultError(NATS_CONNECTION_CLOSED);
+    }
+
+    if (!nc->opts->allowReconnect)
+        return nats_setDefaultError(NATS_INVALID_ARG);
```

Comment:
You need to "unlock". You could also use a setError where you can provide an error text, to be more telling. And I think `NATS_ILLEGAL_STATE` may be more telling?